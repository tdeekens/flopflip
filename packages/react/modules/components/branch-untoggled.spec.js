import React from 'react';
import { shallow } from 'enzyme';
import branchUntoggled from './branch-untoggled';

const UntoggledComponent = () => <div>{'UntoggledComponent'}</div>;
UntoggledComponent.displayName = 'UntoggledComponent';
// This is a shortcut for test expectations on the display name as recompose
// wraps it for us.
UntoggledComponent.wrappedDisplayName = 'renderComponent(UntoggledComponent)';
const FeatureComponent = () => <div>{'FeatureComponent'}</div>;
FeatureComponent.displayName = 'FeatureComponent';

describe('with `flagName`', () => {
  const flagName = 'fooFlagName';

  describe('when feature is enabled', () => {
    const featureFlag = { [flagName]: true };

    describe('without untoggled component', () => {
      let Component;
      let wrapper;

      beforeEach(() => {
        Component = branchUntoggled(undefined, flagName)(FeatureComponent);
        wrapper = shallow(<Component flagName={flagName} {...featureFlag} />);
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should render the `FeatureComponent`', () => {
        expect(wrapper).toRender(FeatureComponent);
      });
    });

    describe('with untoggled component', () => {
      let Component;
      let wrapper;

      beforeEach(() => {
        Component = branchUntoggled(UntoggledComponent, flagName)(
          FeatureComponent
        );
        wrapper = shallow(<Component flagName={flagName} {...featureFlag} />);
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should not render the `UntoggledComponent`', () => {
        expect(wrapper).not.toRender(UntoggledComponent.wrappedDisplayName);
      });

      it('should render the `FeatureComponent`', () => {
        expect(wrapper).toRender(FeatureComponent);
      });
    });
  });
  describe('when feature is disabled', () => {
    const featureFlag = { [flagName]: false };

    describe('with untoggled component', () => {
      let Component;
      let wrapper;

      beforeEach(() => {
        Component = branchUntoggled(UntoggledComponent, flagName)(
          FeatureComponent
        );
        wrapper = shallow(<Component flagName={flagName} {...featureFlag} />);
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should render the `UntoggledComponent`', () => {
        expect(wrapper).toRender(UntoggledComponent.wrappedDisplayName);
      });
    });

    describe('without untoggled component', () => {
      let Component;
      let wrapper;

      beforeEach(() => {
        Component = branchUntoggled(undefined, flagName)(FeatureComponent);
        wrapper = shallow(<Component flagName={flagName} {...featureFlag} />);
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should render `Nothing`', () => {
        expect(wrapper).toRender('Nothing');
      });
    });
  });
});

describe('with `flagName` and `flagVariate`', () => {
  const flagName = 'fooFlagName';
  const flagVariate = 'fooflagVariate';

  describe('when feature is enabled', () => {
    const featureFlag = { [flagName]: flagVariate };

    describe('without untoggled component', () => {
      let Component;
      let wrapper;

      beforeEach(() => {
        Component = branchUntoggled(undefined, flagName, flagVariate)(
          FeatureComponent
        );
        wrapper = shallow(<Component flagName={flagName} {...featureFlag} />);
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should render the `FeatureComponent`', () => {
        expect(wrapper).toRender(FeatureComponent);
      });
    });

    describe('with untoggled component', () => {
      let Component;
      let wrapper;

      beforeEach(() => {
        Component = branchUntoggled(UntoggledComponent, flagName, flagVariate)(
          FeatureComponent
        );
        wrapper = shallow(<Component flagName={flagName} {...featureFlag} />);
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should not render the `UntoggledComponent`', () => {
        expect(wrapper).not.toRender(UntoggledComponent.wrappedDisplayName);
      });

      it('should render the `FeatureComponent`', () => {
        expect(wrapper).toRender(FeatureComponent);
      });
    });
  });

  describe('when feature is disabled', () => {
    const featureFlag = { [flagName]: 'flagVariate2' };

    describe('with untoggled component', () => {
      let Component;
      let wrapper;

      beforeEach(() => {
        Component = branchUntoggled(UntoggledComponent, flagName, flagVariate)(
          FeatureComponent
        );
        wrapper = shallow(<Component flagName={flagName} {...featureFlag} />);
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should render the `UntoggledComponent`', () => {
        expect(wrapper).toRender(UntoggledComponent.wrappedDisplayName);
      });
    });

    describe('without untoggled component', () => {
      let Component;
      let wrapper;

      beforeEach(() => {
        Component = branchUntoggled(undefined, flagName, flagVariate)(
          FeatureComponent
        );
        wrapper = shallow(<Component flagName={flagName} {...featureFlag} />);
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should render `Nothing`', () => {
        expect(wrapper).toRender('Nothing');
      });
    });
  });
});

describe('without `flagName`', () => {
  describe('when feature is enabled', () => {
    const featureFlag = { isFeatureEnabled: true };

    describe('without untoggled component', () => {
      let Component;
      let wrapper;

      beforeEach(() => {
        Component = branchUntoggled()(FeatureComponent);
        wrapper = shallow(<Component {...featureFlag} />);
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should render the `FeatureComponent`', () => {
        expect(wrapper).toRender(FeatureComponent);
      });
    });

    describe('with untoggled component', () => {
      let Component;
      let wrapper;

      beforeEach(() => {
        Component = branchUntoggled(UntoggledComponent)(FeatureComponent);
        wrapper = shallow(<Component {...featureFlag} />);
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should not render the `UntoggledComponent`', () => {
        expect(wrapper).not.toRender(UntoggledComponent.wrappedDisplayName);
      });

      it('should render the `FeatureComponent`', () => {
        expect(wrapper).toRender(FeatureComponent);
      });
    });
  });
  describe('when feature is disabled', () => {
    const featureFlag = { isFeatureEnabled: false };

    describe('with untoggled component', () => {
      let Component;
      let wrapper;

      beforeEach(() => {
        Component = branchUntoggled(UntoggledComponent)(FeatureComponent);
        wrapper = shallow(<Component {...featureFlag} />);
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should render the `UntoggledComponent`', () => {
        expect(wrapper).toRender(UntoggledComponent.wrappedDisplayName);
      });
    });

    describe('without untoggled component', () => {
      let Component;
      let wrapper;

      beforeEach(() => {
        Component = branchUntoggled()(FeatureComponent);
        wrapper = shallow(<Component {...featureFlag} />);
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should render `Nothing`', () => {
        expect(wrapper).toRender('Nothing');
      });
    });
  });

  describe('with default variate value', () => {
    const flagName = 'fooTestFlag';
    const flagVariate = 'fooTestVariate';

    describe('when flag is not defined', () => {
      const featureFlag = {};
      describe('with flag variate and matching default', () => {
        const flagVariateDefaultValue = flagVariate;

        let Component;
        let wrapper;

        beforeEach(() => {
          Component = branchUntoggled(
            undefined,
            flagName,
            flagVariate,
            flagVariateDefaultValue
          )(FeatureComponent);
          wrapper = shallow(<Component flagName={flagName} {...featureFlag} />);
        });

        it('should match snapshot', () => {
          expect(wrapper).toMatchSnapshot();
        });

        it('should render the `FeatureComponent`', () => {
          expect(wrapper).toRender(FeatureComponent);
        });
      });

      describe('with flag variate and non matching default', () => {
        const flagVariateDefaultValue = `${flagVariate}2`;

        let Component;
        let wrapper;

        beforeEach(() => {
          Component = branchUntoggled(
            undefined,
            flagName,
            flagVariate,
            flagVariateDefaultValue
          )(FeatureComponent);
          wrapper = shallow(<Component flagName={flagName} {...featureFlag} />);
        });

        it('should match snapshot', () => {
          expect(wrapper).toMatchSnapshot();
        });

        it('should render the `Nothing`', () => {
          expect(wrapper).toRender('Nothing');
        });
      });
    });

    describe('when flag is defined', () => {
      const featureFlag = { [flagName]: flagVariate };

      describe('with flag variate and matching default', () => {
        const flagVariateDefaultValue = flagVariate;

        let Component;
        let wrapper;

        beforeEach(() => {
          Component = branchUntoggled(
            undefined,
            flagName,
            flagVariate,
            flagVariateDefaultValue
          )(FeatureComponent);
          wrapper = shallow(<Component flagName={flagName} {...featureFlag} />);
        });

        it('should match snapshot', () => {
          expect(wrapper).toMatchSnapshot();
        });

        it('should render the `FeatureComponent`', () => {
          expect(wrapper).toRender(FeatureComponent);
        });
      });
    });
  });
});
