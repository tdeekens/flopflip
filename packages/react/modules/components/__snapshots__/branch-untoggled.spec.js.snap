// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with \`flagName\` and \`flagVariate\` when feature is disabled with untoggled component should match snapshot 1`] = `
<renderComponent(UntoggledComponent)
  flagName="fooFlagName"
  fooFlagName="flagVariate2"
/>
`;

exports[`with \`flagName\` and \`flagVariate\` when feature is disabled without untoggled component should match snapshot 1`] = `
<Nothing
  flagName="fooFlagName"
  fooFlagName="flagVariate2"
/>
`;

exports[`with \`flagName\` and \`flagVariate\` when feature is enabled with untoggled component should match snapshot 1`] = `
<FeatureComponent
  flagName="fooFlagName"
  fooFlagName="fooflagVariate"
/>
`;

exports[`with \`flagName\` and \`flagVariate\` when feature is enabled without untoggled component should match snapshot 1`] = `
<FeatureComponent
  flagName="fooFlagName"
  fooFlagName="fooflagVariate"
/>
`;

exports[`with \`flagName\` when feature is disabled with untoggled component should match snapshot 1`] = `
<renderComponent(UntoggledComponent)
  flagName="fooFlagName"
  fooFlagName={false}
/>
`;

exports[`with \`flagName\` when feature is disabled without untoggled component should match snapshot 1`] = `
<Nothing
  flagName="fooFlagName"
  fooFlagName={false}
/>
`;

exports[`with \`flagName\` when feature is enabled with untoggled component should match snapshot 1`] = `
<FeatureComponent
  flagName="fooFlagName"
  fooFlagName={true}
/>
`;

exports[`with \`flagName\` when feature is enabled without untoggled component should match snapshot 1`] = `
<FeatureComponent
  flagName="fooFlagName"
  fooFlagName={true}
/>
`;

exports[`without \`flagName\` when feature is disabled with untoggled component should match snapshot 1`] = `
<renderComponent(UntoggledComponent)
  isFeatureEnabled={false}
/>
`;

exports[`without \`flagName\` when feature is disabled without untoggled component should match snapshot 1`] = `
<Nothing
  isFeatureEnabled={false}
/>
`;

exports[`without \`flagName\` when feature is enabled with untoggled component should match snapshot 1`] = `
<FeatureComponent
  isFeatureEnabled={true}
/>
`;

exports[`without \`flagName\` when feature is enabled without untoggled component should match snapshot 1`] = `
<FeatureComponent
  isFeatureEnabled={true}
/>
`;

exports[`without \`flagName\` with default variate value when flag is defined with flag variate and matching default should match snapshot 1`] = `
<FeatureComponent
  flagName="fooTestFlag"
  fooTestFlag="fooTestVariate"
/>
`;

exports[`without \`flagName\` with default variate value when flag is not defined with flag variate and matching default should match snapshot 1`] = `
<FeatureComponent
  flagName="fooTestFlag"
/>
`;

exports[`without \`flagName\` with default variate value when flag is not defined with flag variate and non matching default should match snapshot 1`] = `
<Nothing
  flagName="fooTestFlag"
/>
`;
